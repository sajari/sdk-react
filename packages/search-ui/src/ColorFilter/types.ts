import { FilterBoxProps } from '../FilterBox/types';

export interface ColorFilterProps {
  name: string;
  title: FilterBoxProps['title'];
}
