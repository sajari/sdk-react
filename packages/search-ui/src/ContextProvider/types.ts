import { SearchProviderValues } from '@sajari/react-hooks';
import { ThemeProviderProps } from '@sajari/react-sdk-utils';

export interface SearchUIContextProviderValues {
  ratingMax?: number;
  disableDefaultStyles?: boolean;
  customClassNames?: {
    pagination?: {
      container?: string;
      button?: string;
      active?: string;
      disabled?: string;
      pressed?: string;
      next?: string;
      prev?: string;
      spacerEllipsis?: string;
    };
  };
}

export interface ContextProviderValues
  extends SearchProviderValues,
    ThemeProviderProps,
    SearchUIContextProviderValues {}
