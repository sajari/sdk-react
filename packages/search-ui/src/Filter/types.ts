import React from 'react';

export interface HeaderProps {
  title: React.ReactNode;
  showReset?: boolean;
  onReset?: () => void;
}

export interface BoxProps extends HeaderProps {
  children: React.ReactNode;
}

interface BaseFilterProps {
  name: string;
  title: BoxProps['title'];
}

export interface ListFilterProps extends BaseFilterProps {
  type: 'list';
  itemRender?: (value: string) => React.ReactNode;
  /** Maxium of shown items when the list is collapsed */
  limit?: number;
  /** If true, display an input for searching through filter items */
  searchable?: boolean;
  /** The placeholder for search input */
  placeholder?: string;
  /** If true, sort selected items on top */
  pinSelected?: boolean;
  /** How to sort the items in the list */
  sort?: 'count' | 'alpha' | 'none';
  /** Sort in ascending order */
  sortAscending?: boolean;
}

export interface ColorFilterProps extends BaseFilterProps {
  type: 'color';
}

export type FilterProps = ListFilterProps | ColorFilterProps;
