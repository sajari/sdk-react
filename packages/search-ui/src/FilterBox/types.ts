import React from 'react';

export interface HeaderProps {
  title: React.ReactNode;
  showReset?: boolean;
  onReset?: () => void;
}

export interface FilterBoxProps extends HeaderProps {
  children: React.ReactNode;
}
