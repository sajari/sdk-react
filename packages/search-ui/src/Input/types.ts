import { ComboboxProps } from '@sajari/react-components';

export interface InputProps<T> extends Omit<ComboboxProps<T>, 'items' | 'renderItem'> {}
