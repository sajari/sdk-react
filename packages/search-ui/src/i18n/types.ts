export type Locale = {
  common: {
    /** Shown in Summary component e.g 1 result for "television"  */
    result: string;
    /** Shown in Summary component e.g 4 results for "television"  */
    results: string;
    /** Shown when query is empty e.g 4 item  */
    item: string;
    /** Shown when query is empty e.g 4 items */
    items: string;
    /** Shown when performing search */
    loading: string;
    /** Fallback error title when no specific error title is found */
    error: string;
  };
  errors: {
    /** Shown when a search request can't be made due to incorrect pipeline or domain configuration */
    authorization: string;
    /** Shown when there is no network connection */
    connection: string;
    /** Shown when there is an error parsing the result template */
    template: {
      title: string;
      body: string;
    };
    /** Fallback error message when no specific error message is found */
    generic: string;
  };
  filter: {
    /** Shown in TabFilter component as the first tab */
    all: string;
    /** Shown when formatting price */
    rangeOver: string;
    /** Shown when formatting price */
    rangeUnder: string;
    /** Shown as the reset button of a filter */
    reset: string;
    /** Used in SelectFilter, shown as a placeholder when no item is selected */
    select: string;
    /** Used in SelectFilter, show how many items are selected (>= 2) */
    selected: string;
    /** Text to indicate collapsing the filter list */
    showLess: string;
    /** Text to indicate expanding the filter list */
    showMore: string;
    /** Text in button to show filters sidebar  */
    show: string;
    /** Text in button to hide filters sidebar  */
    hide: string;
  };
  input: {
    /** Placeholder for search input */
    placeholder: string;
  };
  resultsPerPage: {
    /** Used in the ResultsPerPage component */
    label: string;
  };
  pagination: {
    /** Used in the Pagination component */
    label: string;
    /** Text used as label for the previous page button */
    previous: string;
    /** Text used as label for the next page button */
    next: string;
    /** Text used as label for a page button */
    page: string;
    /** Text used as label for the current page button */
    current: string;
  };
  results: {
    empty: {
      /** Title shown when no results are found */
      title: string;
      /** Message shown when no results are found */
      body: string;
    };
  };
  result: {
    status: {
      /** The text to show when an product is on sale */
      onSale: string;
      /** The text to show when an product is out of stock */
      outOfStock: string;
      /** The text to show when an product is a new product (within 30 days) */
      newArrival: string;
    };
    /** Label for product's preview image container */
    previewImagesContainer: string;
    /** Label for product's preview image */
    previewImage: string;
  };
  sorting: {
    /** Used in the Sorting component */
    label: string;
    /** "Most Relevant" option */
    mostRelevantOption: string;
  };
  summary: {
    /** The full sentence: {{count}} {{object}} for "{{query}}" */
    results: string;
    /** The full sentence: {{count}} {{object}}  */
    noQueryResults: string;
    /** In seconds */
    latency: string;
    /** Used when completion is enabled in Summary component, the full sentence: Search instead for "{{completion}}" */
    alternative: string;
  };
  viewType: {
    /** Label for ViewType component */
    label: string;
    /** Grid view type */
    grid: string;
    /** List view type */
    list: string;
  };
};
