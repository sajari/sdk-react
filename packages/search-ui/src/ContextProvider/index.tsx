import { Pipeline, SearchContextProvider, Variables } from '@sajari/react-hooks';
import { createContext, ThemeProvider } from '@sajari/react-sdk-utils';
import * as React from 'react';
import { I18nextProvider } from 'react-i18next';

import i18n from '../i18n';
import { ContextProviderValues, SearchUIContextProviderValues } from './types';

const [Provider, useSearchUIContext] = createContext<Required<SearchUIContextProviderValues>>({
  strict: true,
  name: 'PipelineContext',
});

const ContextProvider: React.FC<ContextProviderValues> = ({
  children,
  search,
  importantStyles,
  instant,
  searchOnLoad,
  theme,
  ratingMax = 5,
  disableDefaultStyles = false,
  customClassNames = {},
}) => (
  <Provider value={{ ratingMax, disableDefaultStyles, customClassNames }}>
    <SearchContextProvider search={search} instant={instant} searchOnLoad={searchOnLoad}>
      <ThemeProvider theme={theme} importantStyles={importantStyles}>
        <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
      </ThemeProvider>
    </SearchContextProvider>
  </Provider>
);

export default ContextProvider;
export { Pipeline, useSearchUIContext, Variables };
export type { ContextProviderValues };
