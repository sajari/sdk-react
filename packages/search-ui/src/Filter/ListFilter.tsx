import { Box as CoreBox, Button, Checkbox, CheckboxGroup, Combobox, Radio, RadioGroup } from '@sajari/react-components';
import { useFilter, useQuery } from '@sajari/react-hooks';
import { getStylesObject, isBoolean, isEmpty, isSSR, noop, useTheme } from '@sajari/react-sdk-utils';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import tw from 'twin.macro';

import { IconSmallChevronDown, IconSmallChevronUp } from '../assets/icons';
import { useSearchUIContext } from '../ContextProvider';
import Box from './Box';
import { ListFilterProps } from './types';
import { formatLabel, getHeaderId, pinItems, sortItems } from './utils';

const ListFilter = (props: Omit<ListFilterProps, 'type'>) => {
  const {
    name,
    title,
    limit = 10,
    sort = 'count',
    sortAscending = sort !== 'count',
    itemRender,
    placeholder,
    format,
  } = props;

  const filterContainerId = `list-${name}`;
  const { options, reset, setSelected, selected, multi } = useFilter(name);
  // Enable search by default if there's more than the limit
  const { searchable = options.length > limit } = props;
  // By default, pin selected items if the option count is over limit
  let { pinSelected } = props;
  if (!isBoolean(pinSelected)) {
    pinSelected = options.length > limit;
  }

  // Because as the selected list changes the elements are being recreated so the focus is not focusing the correct element anymore
  const [lastFocusedControl, setLastFocusedControl] = React.useState('');
  const [query, setQuery] = React.useState('');
  const { query: q } = useQuery();
  const [shown, setShown] = React.useState(limit);
  const { disableDefaultStyles = false, customClassNames, currency, language } = useSearchUIContext();
  const theme = useTheme();
  const { t } = useTranslation('filter');

  const styles = getStylesObject(
    {
      innerList: [tw`flex items-center justify-between`],
      count: [tw`ml-2 text-xs text-gray-400`],
      checkbox: [tw`text-sm`],
      searchWrapper: [tw`mb-2`],
      toggleButtonWrapper: [tw`mt-1`],
      toggleIcon: [tw`ml-2`, `color: ${theme.color.primary.base}`],
    },
    disableDefaultStyles,
  );

  // Reset internal query on global query change
  React.useEffect(() => {
    setQuery('');
  }, [q]);

  // Reset shown number when filtering via internal or external query change
  React.useEffect(() => {
    setShown(limit);
  }, [query, q]);

  // On user interaction, get the last element being interacted with BEFORE the list is rerendered and focus that element
  React.useEffect(() => {
    if (!isSSR() && pinSelected) {
      const input = document
        .querySelector(`#${filterContainerId}`)
        ?.querySelector(`input[value='${lastFocusedControl}']`) as HTMLInputElement | null;

      input?.focus();
    }
  }, [JSON.stringify(selected)]);

  const Control = multi ? Checkbox : Radio;
  const filtered = searchable ? options.filter((o) => o.label.toLowerCase().includes(query.toLowerCase())) : options;
  const slice = filtered.length > limit;

  const sortedItems = React.useMemo(() => {
    if (!isSSR() && pinSelected) {
      setLastFocusedControl(`${(document.activeElement as HTMLInputElement).value}`);
    }
    let list = filtered;

    if (sort !== 'none') {
      list = sortItems(list, sort === 'count' ? 'count' : 'label', sortAscending);
    }

    if (pinSelected) {
      list = pinItems(list, selected, 'label');
    }

    return list;
  }, [JSON.stringify(filtered), JSON.stringify(selected), pinSelected, sort, sortAscending]);

  const items = slice ? sortedItems.slice(0, shown) : sortedItems;
  const allShown = shown >= filtered.length;
  const Icon = allShown ? IconSmallChevronUp : IconSmallChevronDown;
  const showMore = React.useCallback(() => {
    if (allShown) {
      setShown(limit);
    } else {
      setShown((s) => s + 20);
    }
  }, [allShown]);

  const innerList = React.useMemo(
    () =>
      items.map(({ label, count }) => (
        <CoreBox css={styles.innerList} key={`${label}-${count}`}>
          <Control
            value={label}
            checked={selected.includes(label)}
            onChange={noop}
            css={styles.checkbox}
            disableDefaultStyles={disableDefaultStyles}
          >
            {typeof itemRender === 'function' ? itemRender(label) : formatLabel(label, { format, currency, t })}
          </Control>
          <span css={styles.count}>{count.toLocaleString(language)}</span>
        </CoreBox>
      )),
    [JSON.stringify(items), itemRender, selected],
  );

  const filterSearchOnChange = React.useCallback((value) => {
    setQuery(value || '');
  }, []);

  if (isEmpty(options) && isEmpty(selected)) {
    return null;
  }

  return (
    <Box title={title} name={name} showReset={selected.length > 0 && multi} onReset={reset}>
      {searchable ? (
        <CoreBox css={styles.searchWrapper}>
          <Combobox
            value={query}
            size="sm"
            className={customClassNames.filter?.list?.searchFilter}
            placeholder={placeholder}
            onChange={filterSearchOnChange}
            disableDefaultStyles={disableDefaultStyles}
          />
        </CoreBox>
      ) : null}

      <CoreBox id={filterContainerId} className={customClassNames.filter?.list?.container}>
        {multi ? (
          <CheckboxGroup
            value={selected}
            onChange={setSelected}
            className={customClassNames.filter?.list?.checkboxGroup}
            disableDefaultStyles={disableDefaultStyles}
            aria-labelledby={getHeaderId(name)}
          >
            {innerList}
          </CheckboxGroup>
        ) : (
          <RadioGroup
            value={selected[0]}
            onChange={(e) => setSelected([e.target.value])}
            className={customClassNames.filter?.list?.radioGroup}
            disableDefaultStyles={disableDefaultStyles}
            aria-labelledby={getHeaderId(name)}
          >
            {innerList}
          </RadioGroup>
        )}
      </CoreBox>

      {slice ? (
        <CoreBox css={styles.toggleButtonWrapper}>
          <Button
            appearance="link"
            onClick={showMore}
            aria-controls={filterContainerId}
            size="sm"
            spacing="none"
            className={customClassNames.filter?.list?.toggleButton}
            disableDefaultStyles={disableDefaultStyles}
          >
            {t(allShown ? 'showLess' : 'showMore')}
            <Icon css={styles.toggleIcon} />
          </Button>
        </CoreBox>
      ) : null}
    </Box>
  );
};

export default ListFilter;
export type { ListFilterProps };
