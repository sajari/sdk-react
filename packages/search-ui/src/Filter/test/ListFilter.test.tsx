import { FilterBuilder, NoTracking, Pipeline, SearchProvider } from '@sajari/react-hooks';
import renderer from 'react-test-renderer';

import { Provider as SearchUIContextProvider } from '../../ContextProvider';
import { SearchUIContextProviderValues } from '../../ContextProvider/types';
import ListFilter from '../ListFilter';

const defaultSearchUIProviderValues: Required<SearchUIContextProviderValues> = {
  tracking: new NoTracking(),
  disableDefaultStyles: false,
  currency: 'USD',
  customClassNames: {},
  ratingMax: 0,
  viewType: 'grid',
  setViewType: () => {},
  downshiftEnvironment: null,
};

const defaultSearchProviderValues = {
  pipeline: new Pipeline(
    {
      account: 'test-account',
      collection: 'test-collection',
    },
    'test-pipeline',
  ),
};

describe('ListFilter rendering', () => {
  it('standard render', () => {
    const priceFilter = new FilterBuilder({
      name: 'price',
      options: {
        High: 'price >= 200',
        Mid: 'price >= 50',
        Low: 'price < 50',
      },
      multi: false,
      initial: ['High'],
    });
    const tree = renderer
      .create(
        <SearchUIContextProvider value={defaultSearchUIProviderValues}>
          <SearchProvider
            search={{
              ...defaultSearchProviderValues,
              filters: [priceFilter],
            }}
            initialResponse={}
          >
            <ListFilter name="price" title="Price Filter" />
          </SearchProvider>
        </SearchUIContextProvider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
