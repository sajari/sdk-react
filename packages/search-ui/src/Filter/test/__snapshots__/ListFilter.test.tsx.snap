// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListFilter rendering standard render 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 0.5em;
  line-height: 1;
}

.emotion-1 {
  padding: 0;
  margin: 0;
  --text-opacity: 1;
  color: rgba(159,166,178,var(--text-opacity));
  font-size: 0.75em;
  font-weight: 500;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  text-transform: uppercase;
  line-height: 1.375;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2>:not(template)~:not(template) {
  --space-y-reverse: 0;
  margin-top: calc(0.25em * calc(1 - var(--space-y-reverse)));
  margin-bottom: calc(0.25em * var(--space-y-reverse));
}

<div>
  <div
    className="emotion-0"
  >
    <h2
      className="emotion-1"
      id="filter-price-label"
    >
      Price Filter
    </h2>
  </div>
  <div
    id="list-price"
  >
    <div
      aria-labelledby="filter-price-label"
      className="emotion-2"
      role="radiogroup"
    />
  </div>
</div>
`;
